postgis/mvt.c

```diff
@@ -687,108 +687,99 @@ LWGEOM *mvt_geom_fast(LWGEOM *lwgeom, const GBOX *gbox, uint32_t extent, uint32_
 {
 	AFFINE affine;
 	gridspec grid;
-	LWGEOM *lwgeom_out = NULL;
 	double width = gbox->xmax - gbox->xmin;
 	double height = gbox->ymax - gbox->ymin;
 	double resx = width / extent;
 	double resy = height / extent;
-	double res = (resx < resy ? resx : resy)/2.0;
+	double res = (resx < resy ? resx : resy)/2;
 	double fx = extent / width;
 	double fy = -(extent / height);
 	double buffer_map_xunits = resx * buffer;
-	static int preserve_collapsed = true;
-	POSTGIS_DEBUGF(2, "%s called", __func__);
+	int preserve_collapsed = LW_TRUE;
+	POSTGIS_DEBUG(2, "mvt_geom called");
 
 	/* Short circuit out on EMPTY */
 	if (lwgeom_is_empty(lwgeom))
 		return NULL;
 
 	if (width == 0 || height == 0)
-		elog(ERROR, "%s: bounds width or height cannot be 0", __func__);
+		elog(ERROR, "mvt_geom: bounds width or height cannot be 0");
 
 	if (extent == 0)
-		elog(ERROR, "%s: extent cannot be 0",  __func__);
-
-	/* Convert any generic GEOMETRYCOLLECTION into a simple */
-	/* base type that reflects one of the contained types */
-	lwgeom_to_basic_type(lwgeom);
+		elog(ERROR, "mvt_geom: extent cannot be 0");
 
 	/* Remove all non-essential points (under the output resolution) */
 	lwgeom_remove_repeated_points_in_place(lwgeom, res);
 	lwgeom_simplify_in_place(lwgeom, res, preserve_collapsed);
+
 	/* If geometry has disappeared, you're done */
 	if (lwgeom_is_empty(lwgeom))
 		return NULL;
 
 	if (clip_geom)
 	{
-		const GBOX *ggbox = lwgeom_get_bbox(lwgeom);
-		GBOX bgbox = *gbox;
+		GBOX bgbox;
+		const GBOX *lwgeom_gbox = lwgeom_get_bbox(lwgeom);;
+		bgbox = *gbox;
 		gbox_expand(&bgbox, buffer_map_xunits);
-		if (!gbox_overlaps_2d(ggbox, &bgbox))
+		if (!gbox_overlaps_2d(lwgeom_gbox, &bgbox))
 		{
-			POSTGIS_DEBUGF(3, "%s: geometry outside clip box", __func__);
+			POSTGIS_DEBUG(3, "mvt_geom: geometry outside clip box");
 			return NULL;
 		}
-		if (!gbox_contains_2d(&bgbox, ggbox))
+		if (!gbox_contains_2d(&bgbox, lwgeom_gbox))
 		{
 			double x0 = bgbox.xmin;
 			double y0 = bgbox.ymin;
 			double x1 = bgbox.xmax;
 			double y1 = bgbox.ymax;
-#if POSTGIS_GEOS_VERSION < 35
-			LWPOLY *lwenv = lwpoly_construct_envelope(0, x0, y0, x1, y1);
-			lwgeom_out = lwgeom_intersection(lwgeom, lwpoly_as_lwgeom(lwenv));
-			lwpoly_free(lwenv);
-#else
-			lwgeom_out = lwgeom_clip_by_rect(lwgeom, x0, y0, x1, y1);
-#endif
-			POSTGIS_DEBUGF(3, "%s: no geometry after clip", __func__);
-			if (lwgeom_out == NULL || lwgeom_is_empty(lwgeom_out))
+			lwgeom = lwgeom_clip_by_rect(lwgeom, x0, y0, x1, y1);
```

Comment:
Is there a problem here w/ the original lwgeom leaking? I guess not in practice, it'll be cleaned up by the memory manager at the end of the call, and there's some small performance benefit in not freeing everything when we know it's going to be cleaned up anyways.

Reply:
I guess that since it's an output function and won't be called per row nor chained it isn't necessary to keep memory under control